### PR TITLE
UCP/WIREUP: Restore compatibility with old UCX versions

### PR DESCRIPTION
**Pull request by dmitrygx**
## What

The intent of the PR is to restore compatibility with old UCX versions.

## Why ?

Since the old UCX versions pack "dedicated - shared" bandwidth value, they could send negative BW, which represents "shared" BW instead of "dedicated" BW which is a positive one.
Fixes the following assertion discovered by MTT:
```
[node01:43706:0:43706]      select.c:270  Assertion `(score >= 0.0) || (rsc_index == ((ucp_rsc_index_t)-1))' failed

/hpc/mtr_scrap/users/mtt/scratch/ucx_ompi/20210908_234502_165745_66027_jazz23.swx.labs.mlnx/installs/uMI4/tests/io_demo/ucx.git/src/ucp/wireup/select.c: [ ucp_wireup_init_select_info() ]
      ...
      267 {
      268     /* score == 0.0 could be specified only when initializing a selection info
      269      * to add CM lane (rsc_index == UCP_NULL_RESOURCE in this case) */
==>   270     ucs_assert((score >= 0.0) || (rsc_index == UCP_NULL_RESOURCE));
      271 
      272     select_info->score      = score;
      273     select_info->addr_index = addr_index;

==== backtrace (tid:  43706) ====
 0 0x000000000015f573 ucp_wireup_init_select_info()  /hpc/mtr_scrap/users/mtt/scratch/ucx_ompi/20210908_234502_165745_66027_jazz23.swx.labs.mlnx/installs/uMI4/tests/io_demo/ucx.git/src/ucp/wireup/select.c:270
 1 0x00000000001610ff ucp_wireup_select_transport()  /hpc/mtr_scrap/users/mtt/scratch/ucx_ompi/20210908_234502_165745_66027_jazz23.swx.labs.mlnx/installs/uMI4/tests/io_demo/ucx.git/src/ucp/wireup/select.c:499
 2 0x0000000000163a58 ucp_wireup_add_bw_lanes()  /hpc/mtr_scrap/users/mtt/scratch/ucx_ompi/20210908_234502_165745_66027_jazz23.swx.labs.mlnx/installs/uMI4/tests/io_demo/ucx.git/src/ucp/wireup/select.c:1135
 3 0x00000000001644ed ucp_wireup_add_rma_bw_lanes()  /hpc/mtr_scrap/users/mtt/scratch/ucx_ompi/20210908_234502_165745_66027_jazz23.swx.labs.mlnx/installs/uMI4/tests/io_demo/ucx.git/src/ucp/wireup/select.c:1379
 4 0x0000000000164fa6 ucp_wireup_search_lanes()  /hpc/mtr_scrap/users/mtt/scratch/ucx_ompi/20210908_234502_165745_66027_jazz23.swx.labs.mlnx/installs/uMI4/tests/io_demo/ucx.git/src/ucp/wireup/select.c:1558
 5 0x000000000016610e ucp_wireup_select_lanes()  /hpc/mtr_scrap/users/mtt/scratch/ucx_ompi/20210908_234502_165745_66027_jazz23.swx.labs.mlnx/installs/uMI4/tests/io_demo/ucx.git/src/ucp/wireup/select.c:1790
 6 0x0000000000171536 ucp_wireup_init_lanes()  /hpc/mtr_scrap/users/mtt/scratch/ucx_ompi/20210908_234502_165745_66027_jazz23.swx.labs.mlnx/installs/uMI4/tests/io_demo/ucx.git/src/ucp/wireup/wireup.c:1315
 7 0x00000000001766be ucp_cm_client_connect_progress()  /hpc/mtr_scrap/users/mtt/scratch/ucx_ompi/20210908_234502_165745_66027_jazz23.swx.labs.mlnx/installs/uMI4/tests/io_demo/ucx.git/src/ucp/wireup/wireup_cm.c:598
 8 0x000000000005654d ucs_callbackq_slow_proxy()  /hpc/mtr_scrap/users/mtt/scratch/ucx_ompi/20210908_234502_165745_66027_jazz23.swx.labs.mlnx/installs/uMI4/tests/io_demo/ucx.git/src/ucs/datastruct/callbackq.c:402
 9 0x0000000000059699 ucs_callbackq_dispatch()  /hpc/mtr_scrap/users/mtt/scratch/ucx_ompi/20210908_234502_165745_66027_jazz23.swx.labs.mlnx/installs/uMI4/tests/io_demo/ucx.git/src/ucs/datastruct/callbackq.h:211
10 0x0000000000065f60 uct_worker_progress()  /hpc/mtr_scrap/users/mtt/scratch/ucx_ompi/20210908_234502_165745_66027_jazz23.swx.labs.mlnx/installs/uMI4/tests/io_demo/ucx.git/src/uct/api/uct.h:2592
11 0x0000000000404d50 UcxContext::progress_worker_event()  /hpc/mtr_scrap/users/mtt/scratch/ucx_ompi/20210908_234502_165745_66027_jazz23.swx.labs.mlnx/installs/uMI4/tests/io_demo/ucx.git/test/apps/iodemo/ucx_wrapper.cc:395
12 0x00000000004043f4 UcxContext::progress()  /hpc/mtr_scrap/users/mtt/scratch/ucx_ompi/20210908_234502_165745_66027_jazz23.swx.labs.mlnx/installs/uMI4/tests/io_demo/ucx.git/test/apps/iodemo/ucx_wrapper.cc:234
13 0x00000000004170ba DemoClient::run()  /hpc/mtr_scrap/users/mtt/scratch/ucx_ompi/20210908_234502_165745_66027_jazz23.swx.labs.mlnx/installs/uMI4/tests/io_demo/ucx.git/test/apps/iodemo/io_demo.cc:1977
14 0x000000000040f21f do_client()  /hpc/mtr_scrap/users/mtt/scratch/ucx_ompi/20210908_234502_165745_66027_jazz23.swx.labs.mlnx/installs/uMI4/tests/io_demo/ucx.git/test/apps/iodemo/io_demo.cc:2614
15 0x000000000040f645 main()  /hpc/mtr_scrap/users/mtt/scratch/ucx_ompi/20210908_234502_165745_66027_jazz23.swx.labs.mlnx/installs/uMI4/tests/io_demo/ucx.git/test/apps/iodemo/io_demo.cc:2657
16 0x00000000000223d5 __libc_start_main()  ???:0
17 0x0000000000403319 _start()  ???:0
=================================
```

## How ?

Use `fabs()` to get absolute value and always consider the received bandwidth value as a dedicated one.

### CODE REVIEW

File: src/ucp/wireup/address.c [hunk 1]
```diff
@@ -505,7 +505,10 @@ ucp_address_unpack_iface_attr(ucp_worker_t *worker,
     packed                = ptr;
     iface_attr->priority  = packed->prio_cap_flags & UCS_MASK(8);
     iface_attr->overhead  = packed->overhead;
-    iface_attr->bandwidth = packed->bandwidth;
+    /* Get only the absolute value of the bandwidth considering the received
+     * bandwidth as a dedicated one to be compatible with the old versions
+     * which could send "dedicated - shared" bandwidth value */
+    iface_attr->bandwidth = fabs(packed->bandwidth);
```
Comment (yosefe) [inline src/ucp/wireup/address.c:511]: many should divide "shared" by ppn here?

Comment (dmitrygx) [reply src/ucp/wireup/address.c:511]: fixed

File: src/ucp/wireup/address.c [hunk 2]
```diff
@@ -502,11 +503,28 @@ ucp_address_unpack_iface_attr(ucp_worker_t *worker,
...
-    iface_attr->lat_ovh   = packed->lat_ovh;
+    packed               = ptr;
+    iface_attr->priority = packed->prio_cap_flags & UCS_MASK(8);
+    iface_attr->overhead = packed->overhead;
+    iface_attr->lat_ovh  = packed->lat_ovh;
+
+    if (packed->bandwidth < 0.0) {
+        /* The received value of the bandwidth is "dedicated - shared" which
+         * doesn't consider ppn and could be sent only by the peer which uses
+         * some old version of UCX. Calculate the bandwidth value considering
```
Comment (yosefe) [inline src/ucp/wireup/address.c:514]: some old version - version v1.11 or order

File: src/ucp/wireup/address.c [hunk 3]
```diff
@@ -502,11 +503,28 @@ ucp_address_unpack_iface_attr(ucp_worker_t *worker,
...
+         * our ppn to be compatible with the peer */
```
Comment (yosefe) [inline src/ucp/wireup/address.c:515]: "our ppn value to be the same value as on the peer"

File: src/ucp/wireup/address.c [hunk 4]
```diff
@@ -502,11 +503,28 @@ ucp_address_unpack_iface_attr(ucp_worker_t *worker,
...
+        bandwidth.shared      = fabs(packed->bandwidth);
+        bandwidth.dedicated   = 0.0;
+        iface_attr->bandwidth = ucp_tl_iface_bandwidth(worker->context,
+                                                       &bandwidth);
+    } else {
+        /* The received value of the bandwidth is either the locally calculated
+         * bandwidth which considers ppn or the dedicated bandwidth sent by the
+         * peer which uses some old version of UCX */
```
Comment (yosefe) [inline src/ucp/wireup/address.c:523]: the received value is either a dedicated bandwidth value (when the peer is using an older version), or the total bandwidth considering remote ppn value (when the peer is using a newer version)

File: src/ucp/wireup/address.c (hunk 2 above)
Comment (dmitrygx) [reply src/ucp/wireup/address.c]: done

File: src/ucp/wireup/address.c (hunk 3 above)
Comment (dmitrygx) [reply src/ucp/wireup/address.c]: done

File: src/ucp/wireup/address.c (hunk 4 above)
Comment (dmitrygx) [reply src/ucp/wireup/address.c]: done

**Review by yosefe — APPROVED** [review]

**Comment by dmitrygx** [comment]
bot:pipe:retest

**Review by brminich — APPROVED** [review]

**Comment by dmitrygx** [comment]
bot:pipe:retest

**Comment by yosefe** [comment]
MellanoxLab test passed http://hpc-master.lab.mtl.com:8080/blue/organizations/jenkins/ucx/detail/ucx/14040/pipeline

**Pull request merged**